sdk/storage/AzureStorageBlob/Source/DataStructures/URLSessionTransferManager.swift

```diff
@@ -0,0 +1,389 @@
+    public static var shared: URLSessionTransferManager = {
+        let shared = URLSessionTransferManager(delegate: nil)
+        return shared
+    }()
+
+    // MARK: TransferManager Methods
+
+    public subscript(index: Int) -> Transfer {
+        // return the operation from the DataStore
+        return transfers[index]
+    }
+
+    // MARK: Add Operations
+
+    public func add(transfer: Transfer) {
+        switch transfer {
+        case let transfer as BlockTransfer:
+            add(transfer: transfer)
+        case let transfer as BlobTransfer:
+            add(transfer: transfer)
+        default:
+            fatalError("Unexpected operation type: \(transfer.self)")
+        }
+    }
+
+    internal func add(transfer: BlockTransfer) {
+        // Add to DataStore
+        transfers.append(transfer)
+
+        // Add to OperationQueue and notify delegate
+        let operation = BlockOperation(withTransfer: transfer)
+        operationQueue.add(operation)
+        self.operation(operation, didChangeState: transfer.state)
+    }
+
+    internal func queueOperationsFor(blobTransfer transfer: BlobTransfer) {
+        let disallowed: [TransferState] = [.complete, .canceled, .failed]
+        guard !disallowed.contains(transfer.state) else { return }
+
+        // Add to OperationQueue and notify delegate
+        let operation = BlobOperation(withTransfer: transfer, queue: operationQueue)
+        var operations: [ResumableTransfer] = [operation]
+        let resumableOperations: [TransferState] = [.pending, .inProgress]
+        let pendingTransfers = transfer.transfers.filter { resumableOperations.contains($0.state) }
+        let initialOperation = BlobInitialOperation(withTransfer: transfer, queue: operationQueue)
+        operations.append(initialOperation)
+        for blockTransfer in pendingTransfers {
+            let blockOperation = BlockOperation(withTransfer: blockTransfer)
+            blockOperation.addDependency(initialOperation)
+            operation.addDependency(blockOperation)
+            operations.append(blockOperation)
+        }
+        operationQueue.add(operations)
+        self.operations(operations, didChangeState: transfer.state)
+    }
+
+    internal func add(transfer: BlobTransfer) {
+        guard let context = persistentContainer?.viewContext else { return }
+
+        // Add to DataStore
+        transfers.append(transfer)
+
+        if transfer.transfers.isEmpty, transfer.state == .pending {
+            let blockTransfer = BlockTransfer.with(context: context, startRange: 0, endRange: 1, parent: transfer)
+            transfer.blocks?.adding(blockTransfer)
+            transfer.totalBlocks = Int64(transfer.transfers.count)
+        }
+        queueOperationsFor(blobTransfer: transfer)
+    }
+
+    // MARK: Cancel Operations
+
+    public func cancel(transfer: Transfer) {
+        transfer.state = .canceled
+        assert(transfer.operation != nil, "Transfer operation unexpectedly nil.")
```

Comment:
Seems like maybe this is not a fatal situation, should we remove the assert? Or do those only occur in debug builds?

Reply:
Yes, the assert will only fail in debug.